src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,202 @@
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(null, value, null);
+        if (size == 0) {
+            head = newNode;
+            tail = newNode;
+            size++;
+            return;
+        }
+        tail.next = newNode;
+        newNode.prev = tail;
+        tail = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        indexCheck(index);
+        Node<T> newNode = new Node<>(null, value, null);
+        if (index == 0) {
+            head.prev = newNode;
+            newNode.next = head;
+            head = newNode;
+            size++;
+            return;
+        }
+        Node<T> nextNode = getNodeByIndex(index);
+        Node<T> prevNode = nextNode.prev;
+        newNode.next = nextNode;
+        newNode.prev = prevNode;
+        prevNode.next = newNode;
+        nextNode.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexCheck(index);
+        Node<T> node = getNodeByIndex(index);
+        return node == null ? null : node.element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        indexCheck(index);
+        Node<T> node = getNodeByIndex(index);
+        if (node == null) {
+            return null;
+        }
+        T oldValue = node.element;
+        node.element = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexCheck(index);
+        Node<T> nodeToRemove = getNodeByIndex(index);
+        return unlink(nodeToRemove);
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> nodeToRemove = head;
+        while (nodeToRemove != null) {
+            if (object == nodeToRemove.element
+                    || object != null && object.equals(nodeToRemove.element)) {
+                break;
+            }
+            nodeToRemove = nodeToRemove.next;
+        }
+        unlink(nodeToRemove);
+        return nodeToRemove == null ? false : true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private T unlink(Node<T> nodeToRemove) {
+        if (nodeToRemove == null) {
+            return null;
+        }
+        if (nodeToRemove.prev == null) {
+            deleteFirstNode();
+            return nodeToRemove.element;
+        }
+        if (nodeToRemove.next == null) {
+            deleteLastNode();
+            return nodeToRemove.element;
+        }
+        deleteNode(nodeToRemove);
+        return nodeToRemove.element;
```

Comment:
Can you simplify this method instead of having a separate logic for each case? delete the first and last nodes very similar